### PR TITLE
Eventbrite block

### PR DESCRIPTION
**Pull request by creativecoder**
#### Changes proposed in this Pull Request:

* Adds `jetpack/eventbrite` block

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?

* New feature, see pb5gDS-aZ-p2

The goal is to release it as a beta block first, in Jetpack 8.2.

#### Testing instructions:

**Inserting the block**

1. Make sure beta blocks are enabled `define( 'JETPACK_BETA_BLOCKS', true );`
1. Insert the Eventbrite block in a page or post and paste an Eventbrite checkout embed code
1. Save the page/post
1. Eventbrite event embed should display in the page/post content
1. Custom eventbrite urls should also work.
1. Test In-page and Modal embed types.

**Transforms**

1. Paste an Eventbrite event url into the block inserter (regular or custom)
1. The block should transform into an Eventbrite block

**Invalid URLs**

1. Insert an Eventbrite block
1. Paste a URL that isn't an Eventbrite event, or an event that doesn't exist
1. You should see an error and an option to convert the block into a link

#### Proposed changelog entry for your changes:

* Block editor: added an Eventbrite block to embed event checkout

### CODE REVIEW

**Comment by creativecoder** [comment]
@pento I'm haven't yet successfully generated the embed output from the `save` function in index.js.

- I originally tried to use `<SandBox>`, though the iframe content always ends up empty, I think because of the [way the content is inserted](https://github.com/WordPress/gutenberg/blob/853b3cb100469f9d71bad93bb38ece5210dae0b2/packages/components/src/sandbox/index.js#L187) with `document.write`, but I haven't verified that.
- Current approach attempts using `dangerouslySetInnerHTML` is not working

(I suppose we could always fall back to a PHP rendered, if necessary)

**Comment by pento** [comment]
We need to render this block in PHP, as the `<script>` tag can't be saved to the post content.

Using a combination of `wp_enqueue_script()` and `wp_add_inline_script()` ensures that `eb_widgets.js` is only loaded once, and that the widgets are initialised after that script is loaded.

**Comment by creativecoder** [comment]
> We need to render this block in PHP, as the `<script>` tag can't be saved to the post content.

Of course 🤦‍♂ !

**Comment by pento** [comment]
I've been playing around with WordPress/gutenberg#19113 today, which would expose Gutenberg's embed block scaffolding, which we could then make use of here.

Of course, Jetpack has back compat requirements, so we can't just start using it, but as this is a new package, I was thinking it would be worth investigating whether we could ship it with Jetpack, and load it on older WordPress versions, where `@wordpress/embed-block` isn't available.

**Comment by creativecoder** [comment]
> I've been playing around with WordPress/gutenberg#19113 today, which would expose Gutenberg's embed block scaffolding, which we could then make use of here.

@pento Given the number of embeds we're working on, that seems like a great idea.

I tried all manner of `npm link`, `yarn link` and `file:../` imports to try and use `@wordpress/embed-block` package from the gutenberg repo in jetpack, but couldn't get any of them to work.

I ended up symlinking `extensions/blocks/embed-block` in jetpack to the `packages/embed-block/src/` folder in the gutenberg repo to get things working locally.

I started with importing the `EmbedPreview` block, but it's not working at the moment (haven't had much time to look into it yet).

**Comment by pento** [comment]
I've updated the PR to use `@wordpress/embed-block`.

Note that, at the moment, this PR requires the Gutenberg plugin to be installed and activated: loading `@wordpress/embed-block` against the block editor in Core causes all sorts of weirdness. 🙂

For running a Docker environment with Jetpack and Gutenberg, I use the Jetpack environment, with a custom `docker-compose.jetpack.yml` file:
```yml
version: '3.3'
services:
  wordpress:
    volumes:
      - /Users/pento/Projects/gutenberg:/var/www/html/wp-content/plugins/gutenberg
```

I start the environment with this command in the Jetpack directory:

`yarn docker:compose -f ../docker-compose.jetpack.yml up -d`

File: extensions/blocks/eventbrite/eventbrite.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ * Handle a failing oEmbed proxy request to try embedding as a shortcode.
+ *
+ * @see https://core.trac.wordpress.org/ticket/45447
+ *
+ * @param  WP_HTTP_Response|WP_Error $response The REST Request response.
+ * @param  WP_REST_Server            $handler  ResponseHandler instance (usually WP_REST_Server).
+ * @param  WP_REST_Request           $request  Request used to generate the response.
+ * @return WP_HTTP_Response|object|WP_Error    The REST Request response.
+ */
+function jetpack_filter_oembed_result( $response, $handler, $request ) {
```
Comment (pento) [inline extensions/blocks/eventbrite/eventbrite.php]: This is copied from the Gutenberg plugin, because embeds fail when the plugin isn't enabled, as Core currently only allows proper oEmbeds to use this REST API endpoint.

I've left it in for now, (even though the PR doesn't run without the plugin), so I don't forget and wonder why everything doesn't work sometime later. 🙃

**Comment by creativecoder** [comment]
> For running a Docker environment with Jetpack and Gutenberg, I use the Jetpack environment, with a custom docker-compose.jetpack.yml file:

@pento Thanks for the instructions--this worked for me without issue. (I wasn't thinking about the `@wordpress/embed-block` to `wp.embedBlock` conversion during build--I think that's why I was having so much trouble.)

Here's a quick screen capture of how this works, for posterity.

Eventbrite Block with embed-block

**Comment by creativecoder** [comment]
@davemart-in See above. Moving to the embed block style means we're using the event URL rather than an embed code snippet--my hunch is that this will be easier for users, but wanted to flag for you since it's different than the original design mockup. Options, like to use the checkout embed as a modal or inline, will be in the sidebar.

**Comment by creativecoder** [comment]
@pento I've updated with a Pinterest copy/paste 😄 

Still to do

- `EVENTBRITE_URL_REGEX` and `URL_REGEX` are not tested
- We need an Eventbrite icon
- The embed preview in the block
- Figure out modal button options

**Comment by pento** [comment]
Thanks for the start, @creativecoder! Here's what I got done today:

- Added an icon.
- As you suggested, renamed the block to "Eventbrite Tickets". Perhaps "Eventbrite Checkout" would be better, since tickets is kind of ambiguous for a service that sells tickets?
- Previews in the editor now work.
- The widget now displays on the published page.

Some issues I didn't get to address during my day:

- Switching the widget type between Modal and Inline doesn't resize the block correctly.
- Something is causing a scrollbar in the Inline preview.
- The Modal preview has no styling.
- `cannotEmbed` should be set by testing against the URL regex.

**Comment by creativecoder** [comment]
@pento

I've refactored the edit component a bit--was finding it a little hard to follow the conditionals over so many lines in the `render` function, hopefully this will help.

I also added the resolve redirect code back in, since [Eventbrite does allow custom urls](https://www.eventbrite.com/support/articles/en_US/How_To/how-to-reuse-a-custom-url-web-address), which don't include the event id.

There are now regexes for the standard Eventbrite and custom Eventbrite URLs. My regex-fu is decent, but not amazing, so I'd appreciate a check of those and any tweaking you recommend. Here's a list of [URLs I've been testing](https://regexr.com/4qurg).

I added a header in the resolve redirect api call, because Eventbrite was returning 403 Forbidden, otherwise.

I left off on trying to figure out why the preview doesn't change when you choose Inline or Modal (until going back to edit the url and reloading the preview). The attributes change but it doesn't trigger a re-render.

(And I just noticed that the edit button in the toolbar shows up in the edit screen with my refactor--need to fix that. I like the `InspectorControls` showing all the time, though, rather than appearing and disappearing)

**Comment by creativecoder** [comment]
@pento Updates for today

- Show edit embed button in toolbar only with the preview
- Show the Embed Type controls all the time (when focused on the block)
- Re-render the preview component when embed type is changed (Here's how the Sandbox component could be changed to re-render when the `html` prop changes https://github.com/Automattic/gutenberg/commit/cfa110b0562ad3eeed46d1dbeca05dcb44af223f, but `key` works in the meantime)
- Fixed issue with preview not showing because transitions were not being hidden

Ideas

- Seems like we should be able to pass a stylesheet to the Sandbox component for styling the modal trigger button, but I didn't get to that
- It would be a good idea to check that Eventbrite returns a successful response for the event url (rather than a 404, for example), otherwise the preview is blank without any indication of what's wrong.

**Comment by creativecoder** [comment]
@pento Today's summary

- Removing interactivity from the preview, since the modal isn't visible anyway, without some resizing hacks for Sandbox
- Updating the resolve-redirect endpoint to return a status, as well, so we can check if the url is actually an event before rendering the preview

I haven't gotten to styling the button preview yet, though I this is probably tied to how we handle editing the button text and any other options we might provide. Since it's essentially a static preview of a button, maybe the Sandbox component isn't actually need?

Also, the modal trigger button displays funny on the front end (at least for twentytwenty), so we'll probably need to tweak the markup there.

**Comment by zinigor** [comment]
It's past code freeze for 8.1, so moving this one to 8.2.

File: extensions/blocks/eventbrite/edit.js [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+				alt: __( 'In page Eventbrite checkout example', 'jetpack' ),
+			},
+			{
+				value: 'modal',
+				isActive: useModal,
+				label: __( ' Button & Modal', 'jetpack' ),
+				image: EventbriteModalExample,
+				alt: __( 'Modal Eventbrite checkout example', 'jetpack' ),
+			},
+		];
```
Comment (Copons) [inline extensions/blocks/eventbrite/edit.js]: The example images are too small when displayed on mobile (at the largest mobile width):

Screenshot 2020-01-16 at 13 55 57

We could either provide larger images (if I'm not mistaken, the container would be ~355px wide), or have them 100% wide in CSS:
```css
.block-editor__container img {
  width: 100%;
}
```

One other thing I've noticed is that the primary button in my in-page embed is green, while in the example is orange.
The button embed too has a different color, but in that case it feels to me less awkward as I can customize it as I wish.

File: extensions/blocks/eventbrite/index.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+			</G>
+		</SVG>
+	),
+	foreground: '#ff8000',
```
Comment (Copons) [inline extensions/blocks/eventbrite/index.js]: I think we should make the icon monochrome to be consistent with other JP blocks, like we did for OpenTable: https://github.com/Automattic/jetpack/pull/14220/commits/327b2ee997175838f57082dd6f2a079deb9cb30b

File: extensions/blocks/eventbrite/modal-button-preview.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		const { borderRadius, placeholder, text, title } = attributes;
+
+		return (
+			<div
+				className={ classnames( className, 'is-modal-button' ) }
+				title={ title }
+				ref={ this.bindRef }
+			>
+				<RichText
+					placeholder={ placeholder || __( 'Add text…' ) }
```
Comment (Copons) [inline extensions/blocks/eventbrite/modal-button-preview.js]: Should we add `jetpack` as domain for all the translated strings in this file?

File: extensions/blocks/eventbrite/edit.js [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+						/>
+						<Button isLarge type="submit">
+							{ _x( 'Embed', 'submit button label', 'jetpack' ) }
+						</Button>
+						{ this.cannotEmbed() && (
+							<p className="components-placeholder__error">
+								{ __( 'Sorry, this content could not be embedded.', 'jetpack' ) }
+								<br />
+								<Button isLarge onClick={ () => fallback( editedUrl, this.props.onReplace ) }>
+									{ _x( 'Convert to link', 'button label', 'jetpack' ) }
```
Comment (Copons) [inline extensions/blocks/eventbrite/edit.js]: The first time I saw this prompt (before looking at the testing steps), I wasn't sure what it actually meant.
What about making it clearer with something like `Convert the block to link`?

File: extensions/blocks/eventbrite/edit.js [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+								</Button>
+							</p>
+						) }
+					</form>
+
+					<div className="components-placeholder__learn-more">
+						<ExternalLink href="http://en.support.wordpress.com/wordpress-editor/blocks/eventbrite-block/">
```
Comment (Copons) [inline extensions/blocks/eventbrite/edit.js]: I assume this is already in our todo, but just in case, let's not forget to add some actual content to this support page.

**Review by Copons — COMMENTED** [review]
Tested and it works as advertised! ✨ 

I've left a few comments on the code, but they're mostly nitpicks.

One thing I'm not sure I got right is the functionality of the "Button & Modal" style.
The name made me think I would have gotten a button opening a modal in my site, while instead it opened an external Eventbrite page (e.g. https://www.eventbrite.com/e/a8c-brighton-test-event-tickets-90126734489#tickets) with a registration modal inside.
Is this the expected behaviour? Is it not working to me because of some oddities going on in my local dev? (e.g. CORS, or some requests blocked by ngrok, etc.)

File: extensions/blocks/eventbrite/edit.js (hunk 5 above)
Comment (creativecoder) [reply extensions/blocks/eventbrite/edit.js]: Great point.

Note that this comes directly from the [Core embed blocks](https://github.com/WordPress/gutenberg/blob/wp/5.3/packages/block-library/src/embed/embed-placeholder.js#L33), so we should submit a change there, as well.

File: extensions/blocks/eventbrite/edit.js (hunk 6 above)
Comment (creativecoder) [reply extensions/blocks/eventbrite/edit.js]: Yes, this is on my todo list--thanks for checking!

File: extensions/blocks/eventbrite/index.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+export const settings = {
```
[the remark this answers is not in the document]
Comment (creativecoder) [reply extensions/blocks/eventbrite/index.js]: This block started out as copy/paste from the Pinterest block, but I agree that following Core is a better idea -- will change.

File: extensions/blocks/eventbrite/index.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	description: __( 'Embed Eventbrite event details and ticket checkout.', 'jetpack' ),
+
+	icon,
+
+	category: 'jetpack',
+
+	keywords: [ __( 'events', 'jetpack' ), __( 'tickets', 'jetpack' ) ],
+
+	supports: {
+		html: false,
```
[the remark this answers is not in the document]
Comment (creativecoder) [reply extensions/blocks/eventbrite/index.js]: > I think we should set this as multiple: false

This is tricky. An earlier version of the block in development had different ids for modal and inline versions, but the Eventbrite widget.js doesn't seem to support multiple embeds for the same event on the same page. But you could have multiple instances of the block as long as it was for different events.

File: extensions/blocks/eventbrite/modal-button-preview.js (hunk 4 above)
Comment (creativecoder) [reply extensions/blocks/eventbrite/modal-button-preview.js]: Good catch. Yes, I think so. But I'm unclear on this--these are all copied directly from Core.

Comment (Copons) [reply extensions/blocks/eventbrite/modal-button-preview.js]: I can see the reason to leave them without domain, but what would happen if Core stops translating those strings?
Honest question, I have no idea! I can guess we'd be left with these strings untranslated, but I'm not sure if there's any mechanism in place for backward compatibility.

@Automattic/i18n maybe y'all know more about this?
For context: in this block we had to recreate some UI _almost_ identical to Core.
The copy is the same so far though, so we added the strings without text domain, in order to reuse the Core's translations.
My fear is that if Core drops those strings, we'd end up with untranslated text.
On the other hand, if we add the `jetpack` domain, we'd need to translate again a bunch of strings that are already translated in Core.
What would you suggest to do?

File: extensions/blocks/eventbrite/edit.js (hunk 5 above)
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: Ah, I see.
I'll try to build a case for changing the string and submit a PR to Guten if it makes sense enough. 👍

File: extensions/blocks/eventbrite/index.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Copons) [reply extensions/blocks/eventbrite/index.js]: Ah, right, that makes perfect sense.
Then we should make the widget ID unique. 🙂

File: extensions/blocks/eventbrite/edit.js (hunk 2 above)
Comment (creativecoder) [reply extensions/blocks/eventbrite/edit.js]: I've updated the image size for the in-page embed and changed the button version to be a kind of live preview, what do you think of that?

File: extensions/blocks/eventbrite/save.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+	const buttonStyle = {
+		backgroundColor: backgroundClass ? undefined : customBackgroundColor,
+		color: textClass ? undefined : customTextColor,
+		borderRadius: borderRadius ? borderRadius + 'px' : undefined,
+	};
+
+	return (
+		<div>
+			<RichText.Content
```
[the remark this answers is not in the document]
Comment (creativecoder) [reply extensions/blocks/eventbrite/save.js]: This is copied directly from how Gutenberg saves the Button block, but using `<button>` instead of `<a>`, since it's an on page action in this context, rather than a link.

It's possible the `<Button>` component could also work, but I think that component is intended for rending a button in the admin view, whereas `<RichText>` seems specifically for rendering the saved block markup that's output in the page content.

File: extensions/blocks/eventbrite/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (creativecoder) [reply extensions/blocks/eventbrite/index.js]: Spacing removed

**Comment by Copons** [comment]
This is ready for JP review @Automattic/jetpack-crew!

I've got an [unresolved i18n doubt](https://github.com/Automattic/jetpack/pull/14075#discussion_r367452517), now hidden under the tons of comments and commits, but it's not a blocker imho.

Also the wpcom diff (D35679-code) seems to be stuck at months ago? 🤔 
Is there a way to resync it manually?

File: extensions/blocks/eventbrite/edit.js (hunk 2 above)
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: I marked this as resolved and then unresolved because otherwise it might get lost in the amount of comments.
I like this solution more, and it's definitely more Gutenberg-y.

There's an issue with it, though: the preview does not automatically resize with the sidebar width.
This can lead to awkward positioning/sizes of the preview, but it's also a Core issue (e.g. the Separator styles suffer the same problem).

You can try it yourself by resizing to a mobile width, opening the sidebar on the Eventbrite block, and then resize the window again (the sidebar should remain open when going to a desktop width).

Just to clarify: I'm just pointing out that this issue exists, but I'm not considering it a blocker.

File: extensions/blocks/eventbrite/eventbrite.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+		"window.EBWidgets.createWidget({
+			widgetType: 'checkout',
+			eventId: ${event_id},
+			modal: true,
+			modalTriggerElementId: '${widget_id}',
+		});"
+	);
+
+	return <<<EOT
+	<noscript><a href="${attr['url']}" rel="noopener noreferrer" target="_blank"></noscript>
```
Comment (jeherve) [inline extensions/blocks/eventbrite/eventbrite.php]: I don't really understand why we need to break the string down into 2 noscript parts, with one part having only `</a>`. Could you explain things a bit?

**Review by jeherve — CHANGES_REQUESTED** [review]
For some reason the button is tiny for me in Twenty Nineteen:

image


--------

I can't seem to be able to focus on the button text to modify its contents:

https://videopress.com/v/7QLN9Ast

(that's in Chrome)

Edit: I was able to click on the button text after a while, I think I was just not clicking in the right place.

-------

It appears that the modal may not be an option on sites without an SSL certificate:

https://videopress.com/v/hTvJJM5x

Maybe we should remove that option when it won't work on the site?

**Comment by Copons** [comment]
> For some reason the button is tiny for me in Twenty Nineteen:
> 
> image
> 
> I can't seem to be able to focus on the button text to modify its contents:
> 
> https://videopress.com/v/7QLN9Ast
> 
> (that's in Chrome)
> 
> Edit: I was able to click on the button text after a while, I think I was just not clicking in the right place.

This is caused by the fact that the button here is an `inline-block` element, while it should in fact be a `block` (see [this comment](https://github.com/WordPress/gutenberg/blob/1f3ea112ba838ef60bf8686326c052ceddf36b4a/packages/block-library/src/file/edit.js#L212) in Guten Core).
👍

File: webpack.config.extensions.js [hunk 11]
```diff
@@ -113,6 +113,10 @@ module.exports = [
 				..._.without( extensionsWebpackConfig.module.rules, transpileConfig ),
 			],
 		},
+		resolve: {
```
[the remark this answers is not in the document]
Comment (Copons) [reply webpack.config.extensions.js]: cc @pento

File: extensions/blocks/eventbrite/eventbrite.php [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+	wp_enqueue_script( 'eventbrite-widget', 'https://www.eventbrite.com/static/widgets/eb_widgets.js', array(), JETPACK__VERSION, true );
+
+	// Show the embedded version.
+	if ( empty( $attr['useModal'] ) ) {
+		wp_add_inline_script(
+			'eventbrite-widget',
+			"window.EBWidgets.createWidget({
+				widgetType: 'checkout',
+				eventId: ${event_id},
```
[the remark this answers is not in the document]
Comment (Copons) [reply extensions/blocks/eventbrite/eventbrite.php]: Do you have any suggestions for the correct function to use?
Most of the `wp_add_inline_script` occurrences in Core escape with `wp_json_encode`, but I've seen some `wp_localize_script` escaping with `esc_attr` or `esc_html`. 🤔

File: extensions/blocks/eventbrite/edit.js [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+		// Check if an Eventbrite URL has been entered, so we need to resolve it.
+		if ( ! prevState.resolvingUrl && this.state.resolvingUrl ) {
+			this.resolveUrl();
+		}
+	}
+
+	// TODO: figure out how to cancel request since apiFetch was updated to use Promises rather than XHR requests.
+	// componentWillUnmount() {}
+
+	resolveUrl = () => {
```
[the remark this answers is not in the document]
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: I think it makes sense, but what about leaving it for a follow up?
I'd rather avoid testing for regressions on a separate block. 🙂

File: extensions/blocks/eventbrite/edit.js [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+import ModalButtonPreview from './modal-button-preview';
+import EventbriteInPageExample from './eventbrite-in-page-example.png';
+import './editor.scss';
+
+const MODAL_BUTTON_STYLES = [
+	{ name: 'fill', label: __( 'Fill' ), isDefault: true },
+	{ name: 'outline', label: __( 'Outline' ) },
+];
+
+class EventbriteEdit extends Component {
```
Comment (scruffian) [inline extensions/blocks/eventbrite/edit.js]: Is there a reason for using a class component over a functional one and hooks? Probably just a matter of taste, but Gutenberg seems to be moving towards all functional components.

File: extensions/blocks/eventbrite/modal-button-preview.js (hunk 4 above)
Comment (lsl) [reply extensions/blocks/eventbrite/modal-button-preview.js]: Please add the text domain.

> I can see the reason to leave them without domain, but what would happen if Core stops translating those strings?

I still need to read up on how jetpack is handling language packs, from what I know so far and going by core i18n docs I suspect the translations may stop loading.

File: webpack.config.extensions.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (pento) [reply webpack.config.extensions.js]: Originally added so that [this](https://github.com/Automattic/jetpack/pull/14075/commits/c3536a6ccb575b84f7ed76517a33afe4e6a31f40#diff-67c4a00b3bd3c01b617670a5fd26e417R13) worked. I haven't kept up with this PR: if it isn't including anything from `_inc/client` anymore, it can be reverted.

File: extensions/blocks/eventbrite/eventbrite.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (jeherve) [reply extensions/blocks/eventbrite/eventbrite.php]: It depends on the value. We could deal with integers with `absint`, text strings with `esc_js`, urls with `esc_url`.

File: extensions/blocks/eventbrite/edit.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeherve) [reply extensions/blocks/eventbrite/edit.js]: Sure thing, happy to do this in a follow-up PR for sure. In this case do you think you could create an issue so the idea doesn't get lost?

File: extensions/blocks/eventbrite/modal-button-preview.js (hunk 4 above)
Comment (Copons) [reply extensions/blocks/eventbrite/modal-button-preview.js]: @lsl Added already! 😄

File: extensions/blocks/eventbrite/eventbrite.php (hunk 10 above)
Comment (Copons) [reply extensions/blocks/eventbrite/eventbrite.php]: I've just poked this and I think I've figured it out.

This renders the button style, which basically is a `<button>` that calls some JS on click, to open the event checkout as a modal.

Without JS, the intention is to render the same button, but inside a simple `<a>` linked to the event on Eventbrite.

To avoid printing the button markup twice, once for JS and once for `noscript`, we mix it up a bit:
- `<noscript><a href="$event_url"><noscript>` only renders without JS
- `$block_content` renders always
- `<noscript></a></noscript>` only renders without JS

Comment (jeherve) [reply extensions/blocks/eventbrite/eventbrite.php]: That makes sense, thank you!

> To avoid printing the button markup twice,

I wonder if it would be worth sacrificing that to make the whole thing more readable?

File: extensions/blocks/eventbrite/edit.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: Done! https://github.com/Automattic/jetpack/issues/14439

File: extensions/blocks/eventbrite/utils.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+import { createBlock } from '@wordpress/blocks';
+
+/**
+ * Fallback behaviour for unembeddable URLs.
+ * Creates a paragraph block containing a link to the URL, and calls `onReplace`.
+ *
+ * @param {string}   url       The URL that could not be embedded.
+ * @param {Function} onReplace Function to call with the created fallback block.
+ */
+export function convertToLink( url, onReplace ) {
```
[the remark this answers is not in the document]
Comment (Copons) [reply extensions/blocks/eventbrite/utils.js]: Done! https://github.com/Automattic/jetpack/issues/14439

File: extensions/blocks/eventbrite/index.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+	),
+	foreground: '#555d66',
+};
+
+export const settings = {
+	title,
+	description: __( 'Embed Eventbrite event details and ticket checkout.', 'jetpack' ),
+	icon,
+	category: 'jetpack',
+	keywords: [ __( 'events', 'jetpack' ), __( 'tickets', 'jetpack' ) ],
```
[the remark this answers is not in the document]
Comment (huguesvincent) [reply extensions/blocks/eventbrite/index.js]: Nothing jumps out from the list of non-found blocks. Conference works for me though meetup is likely a good fit as well in term of # of occurence

File: extensions/blocks/eventbrite/index.js [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+// https://www.eventbrite.com/e/test-event-tickets-123456789
+// https://www.eventbrite.co.uk/e/test-event-tickets-123456789
+export const URL_REGEX = /^\s*https?:\/\/(?:www\.)?(?:eventbrite\.[a-z.]+)\/e\/[^\/]*?(\d+)\/?\s*$/i;
+
+// Custom eventbrite urls use a subdomain of eventbrite.com
+export const CUSTOM_URL_REGEX = /^\s*https?:\/\/(?:.+\.)?(?:eventbrite\.[a-z.]+)\/?\s*$/i;
+
+export const name = 'eventbrite';
+
+export const title = __( 'Eventbrite Checkout', 'jetpack' );
```
[the remark this answers is not in the document]
Comment (huguesvincent) [reply extensions/blocks/eventbrite/index.js]: That is their official name though for that embed (https://www.eventbrite.com/support/articles/en_US/How_To/how-to-add-eventbrite-s-embedded-checkout-to-your-wordpress-org-site?lg=en_US) so I think we need to respect that.

File: extensions/blocks/eventbrite/eventbrite.php (hunk 10 above)
Comment (Copons) [reply extensions/blocks/eventbrite/eventbrite.php]: The code would be more readable, but the button would be rendered twice when JS is disabled. 😄 

- `$block_content` renders always
- `<noscript><a href="$event_url"> $block_content </a><noscript>` only renders without JS

File: extensions/blocks/eventbrite/edit.js (hunk 14 above)
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: I haven't worked on this since the start, and while I agree with you, I'd argue that we're too close to shipping to justify such large refactor. 🙂

**Comment by Copons** [comment]
I think I addressed all the things! ✨

File: extensions/blocks/eventbrite/edit.js (hunk 14 above)
Comment (scruffian) [reply extensions/blocks/eventbrite/edit.js]: I agree, maybe just something worth keeping in mind for future blocks :)

**Review by jeherve — COMMENTED** [review]
I think 25e0fb9ea537008d550607d00b0964ee2b2c41df may have broken things a bit. I can't seem to get a matching event ID from links like:

https://www.eventbrite.com/e/dex-songwriting-expo-tickets-71249967571?aff=ebdshpfprimarybucket

Does it work for you?

The redirect returns the expected response:
```
{"url":"https:\/\/www.eventbrite.com\/e\/dex-songwriting-expo-tickets-71249967571","status":200}
```

But no event ID gets set as a block attribute after that.

**Comment by Copons** [comment]
Oooh awesome catch @jeherve!
It was caused by the fact that the ID regex `/(\d+)\/?\s*$/` matches groups of numbers at the _end of the string_, and of course fails when the URL has any other chars after the ID, such as query params.

The `resolveUrl` method takes care of stripping the query off the URL.
Before, it wasn't a big deal, because the regex was only used at the end of the whole process, on the final URL.

My update instead used the regex as soon as the URL was inserted in the input field, which is _before_ `resolveUrl`, and I totally missed that the URL was set again inside `resolveUrl` as well.

In 338da38 I've added `eventId` to the `setAttributes` inside `resolveUrl`, and now it works with Eventbrite URLs with query params as well.

File: extensions/blocks/eventbrite/edit.js [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+ * Internal dependencies
+ */
+import { createWidgetId, convertToLink, eventIdFromUrl } from './utils';
+import { CUSTOM_URL_REGEX, icon, URL_REGEX } from '.';
+import ModalButtonPreview from './modal-button-preview';
+import EventbriteInPageExample from './eventbrite-in-page-example.png';
+import './editor.scss';
+
+const MODAL_BUTTON_STYLES = [
+	{ name: 'fill', label: __( 'Fill' ), isDefault: true },
```
Comment (jeherve) [inline extensions/blocks/eventbrite/edit.js]: You'll probably want to make those strings translatable. That's not a blocker though, can be done later.

File: extensions/blocks/eventbrite/edit.js [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+		const { setAttributes } = this.props;
+
+		setAttributes( { useModal: 'modal' === embedType } );
+	};
+
+	renderLoading() {
+		return (
+			<div className="wp-block-embed is-loading">
+				<Spinner />
+				<p>{ __( 'Embedding…' ) }</p>
```
Comment (jeherve) [inline extensions/blocks/eventbrite/edit.js]: This can be translated as well.

File: extensions/blocks/eventbrite/edit.js [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+								<br />
+								<Button isLarge onClick={ () => convertToLink( editedUrl, this.props.onReplace ) }>
+									{ _x( 'Convert block to link', 'button label', 'jetpack' ) }
+								</Button>
+							</p>
+						) }
+					</form>
+
+					<div className="components-placeholder__learn-more">
+						<ExternalLink href="http://en.support.wordpress.com/wordpress-editor/blocks/eventbrite-block/">
```
Comment (jeherve) [inline extensions/blocks/eventbrite/edit.js]: We'll need a conditional here so we can have 2 support links, e.g.
https://github.com/Automattic/jetpack/blob/96d06eeb32cf5bfd72b7807c495e9a0ed15a60f3/extensions/blocks/markdown/index.js#L24

File: extensions/blocks/eventbrite/index.js [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+		customTextColor: {
+			type: 'string',
+		},
+		borderRadius: {
+			type: 'number',
+		},
+	},
+	edit,
+	save,
+	transforms: {
```
Comment (jeherve) [inline extensions/blocks/eventbrite/index.js]: This is not working for me for some reason:

https://videopress.com/v/zm4Bbdrg

File: extensions/blocks/eventbrite/save.js [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+				className={ buttonClasses }
+				style={ buttonStyle }
+				value={ text }
+				type="button"
+			/>
+		</div>
+	);
+}
+
+export default function save( { attributes } ) {
```
Comment (jeherve) [inline extensions/blocks/eventbrite/save.js]: I think it would be nice to return a link to the event inside the div / button. Right now, the empty container won't be of much help in subscription emails or if Jetpack gets deactivated.

```html
<div class="wp-block-jetpack-eventbrite"><button id="eventbrite-widget-88665223069" class="wp-block-button__link" type="button">Test one two</button></div>
```

```html
<div id="eventbrite-widget-88665223069" class="wp-block-jetpack-eventbrite"></div>
```

**Review by jeherve — APPROVED** [review]
I have a few last comments, but I don't think any of those are blockers for a first release. Let's try to address those in follow-up PRs?

File: extensions/blocks/eventbrite/edit.js (hunk 18 above)
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: Tracked in https://github.com/Automattic/jetpack/issues/14458

File: extensions/blocks/eventbrite/edit.js (hunk 19 above)
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: Tracked in https://github.com/Automattic/jetpack/issues/14458

File: extensions/blocks/eventbrite/index.js (hunk 21 above)
Comment (Copons) [reply extensions/blocks/eventbrite/index.js]: Tracked by https://github.com/Automattic/jetpack/issues/14460

File: extensions/blocks/eventbrite/edit.js (hunk 20 above)
Comment (Copons) [reply extensions/blocks/eventbrite/edit.js]: Tracked by https://github.com/Automattic/jetpack/issues/14459

**Comment by jeherve** [comment]
Ran into another issue when testing:

#14461

**Comment by jeherve** [comment]
r202093-wpcom

File: extensions/blocks/eventbrite/save.js (hunk 22 above)
Comment (Copons) [reply extensions/blocks/eventbrite/save.js]: Tracked by https://github.com/Automattic/jetpack/issues/14462

File: extensions/blocks/eventbrite/edit.js (hunk 14 above)
Comment (creativecoder) [reply extensions/blocks/eventbrite/edit.js]: This started out as a direct copy and paste from the Pinterest block, that's the only reason--will definitely keep functional components in mind for the next block.

File: extensions/blocks/eventbrite/save.js [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+import { RichText, getColorClassName } from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import { createWidgetId, eventIdFromUrl } from './utils';
+
+/**
+ * Adapted button save function from @wordpress/block-library
+ * (Using Gutenberg code that shipped with WordPress 5.3)
```
[the remark this answers is not in the document]
Comment (creativecoder) [reply extensions/blocks/eventbrite/save.js]: Note that we wouldn't want to remove this. It's different than the button block because it doesn't use a link, but instead triggers an on page modal.
